pipeline-scripts/slacklib.groovy

```diff
@@ -0,0 +1,226 @@
+        }
+        attachments << [
+                title: "Job: <${getBuildURL()}console|#${getBuildNumber()}> by ${owner}",
+                color: '#439FE0',
+        ]
+    }
+
+    withCredentials([string(credentialsId: 'art-bot-slack-token', variable: 'SLACK_BOT_TOKEN')]) {
+
+        base = [   channel: channel,
+                   icon_emoji: ":robot_face:",
+                   username: as_user,
+                   attachments: attachments,
+                   link_names: 1,
+                   reply_broadcast: replyBroadcast,
+                   // icon_url
+        ]
+
+        if ( text ) {
+            base['text'] = text
+        }
+
+        if ( thread_ts ) {
+            base['thread_ts'] = thread_ts
+        }
+
+        def payload = JsonOutput.toJson(base)
+        // echo "Sending slack notification: ${payload}\n"
+        response = httpRequest(
+                        // consoleLogResponseBody: true, // Great for debug, but noisy otherwise
+                        httpMode: 'POST',
+                        quiet: true,
+                        contentType: 'APPLICATION_JSON',
+                        customHeaders: [
+                            [   maskValue: true,
+                                name: 'Authorization',
+                                value: "Bearer $SLACK_BOT_TOKEN"
+                            ]
+                        ],
+                        ignoreSslErrors: true,
+                        requestBody: "${payload}",
+                        url: 'https://slack.com/api/chat.postMessage'
+        )
+
+        if ( verbose ) {
+            echo "[SLACK]>>>\nRequest: ${payload}\nResponse: ${response}\n${response.content}\n<<<[SLACK]\n"
+        }
+
+         //print "Received slack response: ${response}\n\n"
+         return readJSON(text: response.content)
+    }
+}
+
+/**
+ * @param channel_or_release An explicit channel name ('#art-team') or a string that contains a prefix of the
+ *        release the jobs is associated with (e.g. '4.5.2-something' => '4.5'). If a release is specified, the
+ *        slack channel will be #art-release-MAJOR-MINOR.
+ * @param verbose Whether to print the entire slack request and response (optional).
+ * @param as_user Slack handle to submit the message (optional).
+ */
+def to(channel_or_release, verbose=false, as_user='art-release-bot') {
+    def channel = channel_or_release
+    if ( channel_or_release.charAt(0) != '#' ) {
+        channel = getDefaultChannel()
+        if ( channel_or_release.charAt(0).isDigit() ) { // something like 4.5?
+            try {
+                def parts = channel_or_release.split('\\D')  // split on any non-digit; '4.2.5' => ['4','2','5']
+                if ( parts[0] == '3' || parts[0] == '4' ) {
+                    // Turn into a release channel name: e.g. "4.1" -> "#art-release-4-1"
+                    channel = "#art-release-${parts[0]}-${parts[1]}"
```

Comment:
do we need to pre-create these or do they spring into existence?

Reply:
I've already created them. Will invite if we merge. Didn't want to jump the gun if the PR was not what you wanted.